File: app/i18n.py
"""Sprachkataloge: aus einem Schluessel wird Text in der gewuenschten Sprache.

Die Kataloge liegen unter app/static/i18n/ und haben genau zwei Leser: der Browser
holt sie als statische Datei, Python liest dieselbe Datei von der Platte. Eine Datei
je Sprache, zwei Verbraucher, keine zweite Fassung - sonst laufen Oberflaeche und
PDF frueher oder spaeter auseinander.

Platzhalter heissen `{name}`. Diese Schreibweise ist bewusst gewaehlt: derselbe
Ausdruck laesst sich in JavaScript mit einer einzigen Zeile ersetzen, sodass Server
und Oberflaeche denselben Katalogtext identisch fuellen.

Ein fehlender Schluessel darf nie abbrechen: erst wird auf DEFAULT_LOCALE
ausgewichen, und fehlt er auch dort, kommt der Schluessel selbst zurueck. Damit
faellt die Luecke im Bildschirm auf, statt den Ablauf zu toeten.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from functools import lru_cache

from app import config

# Nur `{name}` wird ersetzt. Unbekannte Namen bleiben woertlich stehen - sichtbar,
# aber harmlos; ein KeyError mitten im Fehlertext waere das schlechtere Ergebnis.
_PLACEHOLDER = re.compile(r"\{(\w+)\}")


@dataclass(frozen=True)
class Phrase:
    """Ein Satzbaustein, der erst beim Anzeigen uebersetzt wird.

    Gebraucht, wo ein Text einen zweiten Text einbettet und die Auswahl schon beim
    Rechnen faellt, die Sprache aber erst am Rand feststeht (Beispiel: der
    Aufloesungs-Vorschlag in "output_too_large"). Beim Rendern wird daraus ein
    gewoehnlicher String, sodass die HTTP-Antwort nur flache Werte traegt.
    """

    key: str
    params: dict[str, object] = field(default_factory=dict)


@lru_cache(maxsize=None)
def catalogue(locale: str) -> dict[str, str]:
    """Flachgeklopfter Katalog einer Sprache: "ui.header.title" -> Text.

    Verschachteltes JSON liest sich besser, die Suche braucht aber den Punktpfad.
    Das Ergebnis wird zwischengespeichert; die Dateien aendern sich zur Laufzeit nicht.
    """
    path = config.LOCALE_DIR / f"{locale}.json"
    with path.open(encoding="utf-8") as stream:
        return _flatten(json.load(stream))


def normalise(locale: str | None) -> str:
    """Auf eine unterstuetzte Sprache abbilden. None, Unsinn und "de-CH" gehen durch."""
    if not locale:
        return config.DEFAULT_LOCALE
    tag = str(locale).strip().lower().replace("_", "-")
    primary = tag.split("-")[0]
    return primary if primary in config.SUPPORTED_LOCALES else config.DEFAULT_LOCALE


def translate(key: str, locale: str, /, **params: object) -> str:
    """Text zu einem Punktschluessel, mit `{name}`-Platzhaltern gefuellt.

    Die Parameter stehen positionell vor dem `/`, damit ein Platzhalter ruhig
    "key" oder "locale" heissen darf.
    """
    text = _lookup(key, locale)
    if text is None:
        return key
    return _substitute(text, {name: _plain(value, locale) for name, value in params.items()})


def plain_params(params: dict[str, object], locale: str) -> dict[str, object]:
    """Parameter fuer die HTTP-Antwort: eingebettete Phrase-Bausteine werden Text."""
    return {name: _plain(value, locale) for name, value in params.items()}


def negotiate(accept_language: str | None) -> str:
    """Beste unterstuetzte Sprache aus einem Accept-Language-Kopf, sonst DEFAULT_LOCALE."""
    if not accept_language:
        return config.DEFAULT_LOCALE

    ranked: list[tuple[float, int, str]] = []
    for index, part in enumerate(accept_language.split(",")):
        tag, _, parameters = part.strip().partition(";")
        tag = tag.strip().lower()
        if tag:
            ranked.append((-_quality(parameters), index, tag))

    # Nach Gewicht, bei Gleichstand in der Reihenfolge des Kopfes - so gewinnt bei
    # "de,en" das zuerst genannte Deutsch.
    for _, _, tag in sorted(ranked):
        if tag == "*":
            return config.DEFAULT_LOCALE
        primary = tag.split("-")[0]
        if primary in config.SUPPORTED_LOCALES:
            return primary
    return config.DEFAULT_LOCALE


def _quality(parameters: str) -> float:
    """Das q-Gewicht eines Accept-Language-Eintrags; ohne Angabe gilt 1.0."""
    for parameter in parameters.split(";"):
        parameter = parameter.strip()
        if parameter.startswith("q="):
            try:
                return float(parameter[2:])
            except ValueError:
                return 0.0
    return 1.0


def _lookup(key: str, locale: str) -> str | None:
    """Schluessel in der Wunschsprache, sonst in der Vorgabesprache, sonst None."""
    wanted = normalise(locale)
    text = catalogue(wanted).get(key)
    if text is None and wanted != config.DEFAULT_LOCALE:
        text = catalogue(config.DEFAULT_LOCALE).get(key)
    return text


def _plain(value: object, locale: str) -> object:
    return translate(value.key, locale, **value.params) if isinstance(value, Phrase) else value


def _substitute(text: str, params: dict[str, object]) -> str:
    return _PLACEHOLDER.sub(
        lambda match: str(params[match.group(1)]) if match.group(1) in params else match.group(0),
        text,
    )


def _flatten(data: dict[str, object], prefix: str = "") -> dict[str, str]:
    flat: dict[str, str] = {}
    for name, value in data.items():
        path = f"{prefix}{name}"
        if isinstance(value, dict):
            flat.update(_flatten(value, f"{path}."))
        else:
            flat[path] = str(value)
    return flat

File: web/pdf/i18n.js
/**
 * i18n.js - aus einem Schluessel wird Text. Invariante 7 fuer den PDF-Bau.
 *
 * Es sind DIESELBEN Katalogdateien, die auch app/i18n.py liest und die die
 * Oberflaeche holt (app/static/i18n/). Keine zweite Fassung - genau deshalb kann
 * tests/test_i18n.py darauf bestehen, dass beide Sprachen denselben Schluesselsatz
 * haben, und genau deshalb steht im PDF kein fertiger deutscher Satz im Code.
 *
 * Warum nicht app/static/js/i18n.js benutzt wird, das es schon gibt: das Modul der
 * Oberflaeche holt die Kataloge ueber `fetch`, merkt sich die gewaehlte Sprache,
 * schreibt `<html lang>` und ist asynchron. Ein PDF-Bau braucht davon nichts und
 * darf nichts davon voraussetzen - er braucht eine reine Funktion. Uebernommen ist
 * deshalb die Fassung aus app/i18n.py, nicht die des Browsers.
 *
 * Ein fehlender Schluessel bricht nie ab: erst wird auf DEFAULT_LOCALE ausgewichen,
 * und fehlt er auch dort, kommt der Schluessel selbst zurueck. Damit faellt die
 * Luecke auf dem Papier auf, statt den Ablauf zu toeten.
 */

import de from "../../app/static/i18n/de.json" with { type: "json" };
import en from "../../app/static/i18n/en.json" with { type: "json" };
import { DEFAULT_LOCALE, SUPPORTED_LOCALES } from "./constants.js";

// Nur `{name}` wird ersetzt. Unbekannte Namen bleiben woertlich stehen - sichtbar,
// aber harmlos; ein Absturz mitten im Fehlertext waere das schlechtere Ergebnis.
const PLACEHOLDER = /\{(\w+)\}/g;

const SOURCES = { de, en };

// Statische Importe kann man nicht aus einer Liste erzeugen. Damit eine neue
// Sprache in shared/constants.json nicht still ohne Katalog bleibt, wird der
// Abgleich hier EINMAL beim Laden gemacht - laut und frueh statt spaeter als
// englischer Satz auf einem deutschen Blatt.
for (const locale of SUPPORTED_LOCALES) {
    if (!(locale in SOURCES)) {
        throw new Error(`Sprache "${locale}" steht in shared/constants.json, aber web/pdf/i18n.js kennt keinen Katalog dafuer.`);
    }
}

const CATALOGUES = Object.fromEntries(
    Object.entries(SOURCES).map(([locale, tree]) => [locale, flatten(tree)]),
);

/** Verschachteltes JSON zu "a.b.c" -> Text. */
function flatten(node, prefix = "", target = {}) {
    for (const [name, value] of Object.entries(node)) {
        const path = `${prefix}${name}`;
        if (value !== null && typeof value === "object") {
            flatten(value, `${path}.`, target);
        } else {
            target[path] = String(value);
        }
    }
    return target;
}

/** Auf eine unterstuetzte Sprache abbilden. null, Unsinn und "de-CH" gehen durch. */
export function normalise(locale) {
    if (!locale) return DEFAULT_LOCALE;
    const primary = String(locale).trim().toLowerCase().replace(/_/g, "-").split("-")[0];
    return SUPPORTED_LOCALES.includes(primary) ? primary : DEFAULT_LOCALE;
}

/** Text zu einem Punktschluessel, mit `{name}`-Platzhaltern gefuellt. */
export function translate(key, locale, params = {}) {
    const wanted = normalise(locale);
    let text = CATALOGUES[wanted][key];
    if (text === undefined && wanted !== DEFAULT_LOCALE) {
        text = CATALOGUES[DEFAULT_LOCALE][key];
    }
    if (text === undefined) return key;
    return text.replace(PLACEHOLDER, (match, name) =>
        Object.hasOwn(params, name) ? String(params[name]) : match,
    );
}
